docs/dev/adding_instruments.rst

```diff
@@ -732,6 +644,102 @@ In the above example, :code:`MultimeterA` and :code:`MultimeterB` use a differen
+
+.. code-block:: python
+
+    class Extreme5002(Instrument):
+        """The great Extreme5002 instrument.
+        :param address: The device address for the communication.
+        """
+        def __init__(self, adapter, address=0):
+            super().__init__(adapter, "Extreme5002")
+            self.address = f"{address:03}"
+
+        def write(self, command):
+            """Add the device address in front of every command and send it to the device."""
+            super().write(self.address + command)
+
+        def read(self):
+            """Read from the device and assert, that the response starts with the device address."""
+            got = super().read()
+            if got.startswith(self.address):
+                return got[3:]
+            else:
+                raise ConnectionError(f"Message {got} for wrong address read.")
+    
+        voltage = Instrument.measurement(
+            ":VOLT:?", "The measured voltage in Volts.")
+
+If the device is initialized with :code:`address=12`, a request for the voltage would send :code:`"012:VOLT:?"` to the device and expect a response beginning with :code:`"012"`.
+
+
+Bytes communication
+*******************
+
+Some devices do not expect ASCII strings but raw bytes. In those cases, you can call the :meth:`write_bytes` and :meth:`read_bytes` in your :meth:`write` and :meth:`read` methods. The following example shows an instrument, which has registers to be written and read via bytes sent.
+
+.. code-block:: python
+
+    class Extreme5003(Instrument):
+    """The Extreme5003 instrument with bytes communication."""
+        def __init__(self, adapter):
+            super().__init__(adapter, "Extreme5003")
+    
+        def write(self, command):
+            """Write to the device according to the comma separated command.
+    
+            :param command: R or W for read or write, hexadecimal address, and data.
+            """
+            function, address, data = command.split(",")
+            b = [0x03] if function == "R" else [0x10]
+            b.append(int(address, 16))  # register address
+            b.extend(int(data).to_bytes(length=8, byteOrder="big", signed=True)
+            self.write_bytes(b)
+    
+        def read(self):
+            """Read the response and return the data as an integer, if applicable."""
+            response = read_bytes(2)  # return type and payload
+            if response[0] == 0x00:
+                raise ConnectionError(f"Device error of type {response[1]} occurred.")
+            if response[0] == 0x03:
+                # read that much bytes and return them as an integer
+                data = self.read_bytes(response[1])
+                return int.from_bytes(data, byteorder="big", signed=True)
+            if response[0] == 0x10 and response[1] != 0x00:
+                raise ConnectionError(f"Writing to the device failed with error {response[1]}")
+    
+        voltage = Instrument.control(
+            "R,0x106,1", "W,0x106,%i",
```

Comment:
I love how concise and simple this is now!

Reply:
Thanks